# B5/B6/B7 — Review variant_claims candidates, guarantee floor conflicts, measure extraction quality

Per `docs/TODO-stage4.md` Track B:
- **B5**: review every row in `variant_claims_candidates.json` against its source
  `passage_ref`; promote approved rows by setting `trust_tier=1`.
- **B6**: guarantee the seed contains all three floor conflicts (Aphrodite parentage,
  Io parentage, Achilles death) regardless of what extraction found — hand-add only
  what extraction missed.
- **B7** (diagnostic, non-blocking): measure how many of the two *measurable* floor
  conflicts (Aphrodite, Achilles — Io is structurally excluded, see below) the pipeline
  detected unaided.

Run B2 (`python -m extraction.run_extraction`) first if you haven't — this notebook
reads its output from `extraction/output/variant_claims_candidates.json` and writes
back to the same file (never a second copy — `trust_tier` is the only thing that
changes on review).

In [12]:
import json
import sys
from pathlib import Path

INGESTION_ROOT = Path.cwd().parent
sys.path.insert(0, str(INGESTION_ROOT))

from dotenv import load_dotenv
load_dotenv()

CANDIDATES_PATH = INGESTION_ROOT / "extraction" / "output" / "variant_claims_candidates.json"

In [13]:
from loader.source_registry import SOURCE_REGISTRY
from loader.text_cleaner import clean
from extraction.segmentation import segment

candidates = json.loads(CANDIDATES_PATH.read_text(encoding="utf-8"))
print(f"{len(candidates)} candidate rows loaded")

7429 candidate rows loaded


## B7 — Extraction-quality metric (diagnostic, non-blocking)

Measured against the raw candidates, **before** any B6 hand-add below. Only Aphrodite
and Achilles are measurable this way (`N/2`, not `N/3`) — Io's two variants are both
attributed to Apollodorus (a single `source_id`), so the `count(DISTINCT source_id) >=
2` detector can never emit it; that's expected, not a pipeline miss.

In [14]:
def floor_conflict_detected(subject: str, claim_type: str) -> bool:
    rows = [
        c for c in candidates
        if c["subject_name"].strip().lower() == subject.lower() and c["claim_type"] == claim_type
    ]
    return len({c["source_id"] for c in rows}) >= 2


aphrodite_detected = floor_conflict_detected("Aphrodite", "parentage")
achilles_detected = floor_conflict_detected("Achilles", "death")
detected_count = sum([aphrodite_detected, achilles_detected])

print(f"floor conflicts detected: {detected_count}/2")
print(f"  Aphrodite parentage: {'detected' if aphrodite_detected else 'MISSED'}")
print(f"  Achilles death:      {'detected' if achilles_detected else 'MISSED'}")
print("  Io parentage:        structurally excluded (single-source, not measurable)")

floor conflicts detected: 2/2
  Aphrodite parentage: detected
  Achilles death:      detected
  Io parentage:        structurally excluded (single-source, not measurable)


## B5 — Review every candidate row, grouped by disagreement

Dumping all 7,429 rows in one cell floods Jupyter's IOPub rate limit (and isn't readable anyway). The real review unit is the **841 distinct `(subject, claim_type)` groups** -- each one disagreement, with 1+ candidate rows supporting each side. Review one group at a time with `review_group(subject, claim_type)` below.

For each row shown: check the `claim_value` against the segment text -- the exact text the LLM saw, re-derived the same way `run_extraction.py` builds it (**not** a `narrative_chunks` lookup, which uses different chunk boundaries) -- and decide whether it's a real, correctly-attributed claim.

In [15]:
def build_segment_map(source):
    raw = (INGESTION_ROOT / source.file_path).read_text(encoding="utf-8")
    cleaned = clean(raw)
    segs = segment(cleaned, source.author, source.work, source.passage_ref_extractor)
    return {s.passage_ref: s.text for s in segs}


segment_maps = {s.source_id: build_segment_map(s) for s in SOURCE_REGISTRY}

In [ ]:
from collections import defaultdict

GROUP_INVENTORY_PATH = INGESTION_ROOT / "audit" / "group_inventory_findings.json"

groups: dict[tuple[str, str], list[int]] = defaultdict(list)
for i, c in enumerate(candidates):
    groups[(c["subject_name"].strip().lower(), c["claim_type"])].append(i)

print(f"{len(groups)} distinct (subject, claim_type) groups across {len(candidates)} candidate rows")
print("(one line per group -- pick a subject/claim_type below and call review_group(...) to see its rows)\n")

# Track C6 (DEV-104): ranked by Track B's A8 subject prominence instead of the old flat
# alphabetical dump, so a tranche is picked by prominence + source count rather than by
# scrolling. Reads A10's JSON artifact from disk -- never imports the audit package, so this
# notebook doesn't grow a dependency on its internals (B7 emits the artifact for exactly this).
# A10 groups by *canonical* claim_type + alias-resolved subject; this notebook's own `groups`
# dict is raw (lowered subject, raw claim_type) -- matching on lowered subject alone is enough
# to carry the subject's rank over onto every raw group it owns.
if GROUP_INVENTORY_PATH.exists():
    inventory = json.loads(GROUP_INVENTORY_PATH.read_text(encoding="utf-8"))["groups"]
    rank_by_subject = {row["subject"].strip().lower(): row["subjectRank"] for row in inventory}
    UNRANKED = float("inf")

    def _sort_key(item):
        (subject, claim_type), idxs = item
        return (rank_by_subject.get(subject, UNRANKED), subject, claim_type)

    print(f"(ranked by A8 subject prominence, from {GROUP_INVENTORY_PATH.name})\n")
    for (subject, claim_type), idxs in sorted(groups.items(), key=_sort_key):
        distinct_sources = len({candidates[i]["source_id"] for i in idxs})
        rank = rank_by_subject.get(subject, UNRANKED)
        rank_label = f"rank {rank:>4}" if rank != UNRANKED else "rank  n/a"
        print(f"  {rank_label}  {subject!r:30} {claim_type!r:20} {len(idxs)} rows, {distinct_sources} distinct sources")
else:
    print(
        f"(no A10 inventory found at {GROUP_INVENTORY_PATH} -- run "
        "`python -m audit.group_inventory --db` first for a prominence-ranked listing; "
        "falling back to alphabetical)\n"
    )
    for (subject, claim_type), idxs in sorted(groups.items()):
        distinct_sources = len({candidates[i]["source_id"] for i in idxs})
        print(f"  {subject!r:30} {claim_type!r:20} {len(idxs)} rows, {distinct_sources} distinct sources")

In [ ]:
from extraction.run_extraction import _claim_key


def review_group(subject: str, claim_type: str) -> None:
    """Prints every candidate row for one (subject, claim_type) group, with its
    source segment text. Call this repeatedly for whichever groups you're reviewing
    right now -- output stays small per call, so no IOPub flooding. Ends with a
    ready-to-paste `approved_keys += [...]` line covering every row shown -- paste
    it below, delete whichever rows you're rejecting, then run.

    Track C2 (DEV-104): keyed by the same 5-tuple identity DEV-101's merge-on-write
    uses (`extraction.run_extraction._claim_key` -- imported, never redefined), not
    by position. `variant_claims_candidates.json` is rewritten in extraction order
    on every re-extraction, so an index recorded in one review session can point at
    an entirely different, unrelated claim by the next session -- proven live, not
    hypothetical (see the markdown cell above the promotion cells below)."""
    key = (subject.strip().lower(), claim_type)
    idxs = groups.get(key)
    if not idxs:
        print(f"no group for {subject!r}/{claim_type!r} -- check the listing above for the exact claim_type spelling")
        return
    row_keys = []
    for i in idxs:
        c = candidates[i]
        row_key = _claim_key(c)
        row_keys.append(row_key)
        print("=" * 80)
        print(f"[{i}] key={row_key}")
        print(f"    subject={c['subject_name']!r}  claim_type={c['claim_type']!r}")
        print(f"    claim_value: {c['claim_value']}")
        print(f"    source_id={c['source_id']}  passage_ref={c['passage_ref']}  trust_tier={c['trust_tier']}")
        segment_text = segment_maps.get(c["source_id"], {}).get(c["passage_ref"])
        if segment_text is None:
            print("    [!] no matching segment for this passage_ref -- verify manually against the source text")
        else:
            # preview = segment_text[:400] + ("..." if len(segment_text) > 400 else "")
            preview = segment_text
            print(f"    segment text: {preview}")
    print("-" * 80)
    print("copy/paste (delete any rows you're rejecting):")
    print(f"approved_keys += {row_keys!r}")


# Example -- start with the floor conflicts:
review_group("Io", "parentage")

## Decide which rows to promote

Edit the list below with the keys (from the `approved_keys += [...]` line printed by
`review_group`) you approve this session -- run it, then repeat across as many sessions as you
like. **Additive only**: this only ever sets rows to `trust_tier=1`, never demotes an
already-promoted row back down, so partial review across multiple sittings is safe -- you don't
need to re-list keys you approved in an earlier session.

**Track C (DEV-104): keyed, not positional.** Promotion here used to be `candidates[i]["trust_tier"]
= 1` against `approved_indices: list[int]` -- unsafe the moment this file is ever regenerated by a
re-extraction, since `variant_claims_candidates.json` is rewritten in extraction order (DEV-101),
so index `i` can silently point at a different claim next time. The cell below proves this
happened, not just that it could.

In [ ]:
# <-- EDIT THIS after reviewing the rows above
# Keyed by (subject_name, claim_type, claim_value, source_id, passage_ref) -- the same 5-tuple
# DEV-101's merge-on-write uses (extraction.run_extraction._CLAIM_IDENTITY) -- never by position.
approved_keys: list[tuple] = []

### Historical promotions, migrated from position to key (Track C1/C3, DEV-104)

**C1 -- the hazard, proven, not assumed.** The three `approved_indices += [...]` cells below used
to record this notebook's early review sessions by position. Replaying `run_extraction.py`'s real
merge-on-write (`_write_claims_preserving_review`) against a reordered copy of the live 7,429-row
file -- same claims, different order, exactly what a re-extraction with a different segment
processing order produces -- showed **73 of 73** of those historical indices now pointing at a
completely different, unrelated claim (e.g. index `743` used to be `('Aphrodite', 'parentage',
'child of Zeus', ...)`; after the reorder it resolves to `('Agamemnon', 'death', 'killed by
Aegisthus', ...)`). No error, no warning -- exactly ADR-004's gate failing silently. DEV-101's
key-based merge itself worked correctly throughout (all 71 promoted **keys** survived the
reorder); the hazard is specific to a notebook that re-derives *positions* across sessions.

**C3 -- the migration found a real drift, not just a hypothetical one.** Resolving the three
historical cells' indices against **today's** live file (no simulation) already disagrees with
what's actually promoted: only **71 of those 73** indices currently resolve to a `trust_tier=1`
row, and one promoted row (`Aphrodite`/`birth`) isn't covered by any of the three cells at all. Per
this track's own instruction ("a mismatch means the file already drifted -- stop and triage, do
not proceed"): the three stale index cells below are **not** used as the migration source. The
cell below instead lists the **71 keys the live file actually has at `trust_tier=1` right now** --
the real, current review decision, which DEV-101's merge has been correctly preserving by key
since it landed -- across exactly the 4 groups the Contracts section names: `Aphrodite/parentage`
(33), `Aphrodite/birth` (1), `Achilles/death` (27), `Io/parentage` (10).

In [ ]:
approved_keys += [
    ('Achilles', 'death', 'Achilles was slain; Ajax carried his dead body from the battlefield', 'ovid-metamorphoses', '13.280-13.312'),
    ('Achilles', 'death', 'Achilles wished Hector had slain him instead of dying swept away by the river', 'homer-iliad', '21.272-21.297'),
    ('Achilles', 'death', 'Foretold by dying Hector to be slain by Paris and Phoebus Apollo at the Scaean gate', 'homer-iliad', '22.328-22.374'),
    ('Achilles', 'death', 'Killed by Paris and Apollo with arrows/shafts', 'ovid-metamorphoses', '13.481-13.507'),
    ('Achilles', 'death', "Killed by an arrow shot by Paris, guided by Apollo, at Apollo's and Neptune's instigation - a divergence from the more common tradition of Achilles being wounded in the heel", 'ovid-metamorphoses', '12.580-12.612'),
    ('Achilles', 'death', "Thetis foretells his death will follow soon after Hector's death", 'homer-iliad', '18.50-18.96'),
    ('Achilles', 'death', 'Thetis foretold he would perish beneath the wall of the Trojans by the swift missiles of Apollo', 'homer-iliad', '21.272-21.297'),
    ('Achilles', 'death', 'Thetis forewarned Achilles that he would die by the hand of Apollo if he slew Tenes', 'apollodorus-bibliotheca', 'E.3.20-E.3.26'),
    ('Achilles', 'death', 'denied by unjust fate to both the Greeks and Ulysses, implying his death before the contest for his arms', 'ovid-metamorphoses', '13.128-13.161'),
    ('Achilles', 'death', 'destined to meet his death and fate at some future point', 'homer-iliad', '20.292-20.339'),
    ('Achilles', 'death', 'destined to perish in deep-soiled Troy, far from his native land', 'homer-iliad', '24.77-24.119'),
    ('Achilles', 'death', 'doomed to a speedy death, brief span of life', 'homer-iliad', '1.386-1.427'),
    ('Achilles', 'death', 'doomed to be brought low beneath the wall of the wealthy Trojans', 'homer-iliad', '23.54-23.98'),
    ('Achilles', 'death', 'fated to be slain in fight by a god and a mortal, far from his father and mother', 'homer-iliad', '19.392-20.11'),
    ('Achilles', 'death', 'foretold that he will die in battle, smitten by a spear cast or an arrow from the string', 'homer-iliad', '21.97-21.135'),
    ('Achilles', 'death', 'killed by Alexander', 'apollodorus-bibliotheca', 'E.4.6-E.5.4'),
    ('Achilles', 'death', 'killed by Apollo', 'apollodorus-bibliotheca', 'E.3.20-E.3.26'),
    ('Achilles', 'death', 'killed by Apollo', 'apollodorus-bibliotheca', 'E.4.6-E.5.4'),
    ('Achilles', 'death', 'killed by Apollo', 'homer-iliad', '19.392-20.11'),
    ('Achilles', 'death', 'killed by Apollo', 'homer-iliad', '21.272-21.297'),
    ('Achilles', 'death', 'killed by Apollo', 'homer-iliad', '22.328-22.374'),
    ('Achilles', 'death', 'killed by Apollo', 'ovid-metamorphoses', '13.481-13.507'),
    ('Achilles', 'death', 'killed by Paris', 'ovid-metamorphoses', '12.580-12.612'),
    ('Achilles', 'death', 'killed by Paris', 'ovid-metamorphoses', '13.162-13.204'),
    ('Achilles', 'death', 'killed by Paris', 'ovid-metamorphoses', '13.481-13.507'),
    ('Achilles', 'death', 'slain at Troy', 'homer-odyssey', '3.102-3.154'),
    ('Achilles', 'death', 'slain in the land of Troy, far from Argos', 'homer-odyssey', '24.35-24.84'),
    ('Aphrodite', 'birth', 'Born from the sea foam formed by the severed members of Heaven, cast into the sea by Cronus', 'hesiod-theogony', '176-232'),
    ('Aphrodite', 'parentage', 'child of Dione', 'apollodorus-bibliotheca', '1.3.1-1.3.5'),
    ('Aphrodite', 'parentage', 'child of Dione', 'homer-iliad', '5.334-5.380'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'apollodorus-bibliotheca', '1.3.1-1.3.5'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'hesiod-homeric-hymns', '3.182-3.238'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'hesiod-homeric-hymns', '5.106-5.154'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'hesiod-homeric-hymns', '5.53-5.105'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '14.153-14.196'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '14.197-14.241'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '20.103-20.143'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '3.340-3.380'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '5.297-5.333'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '5.334-5.380'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '5.778-5.824'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-odyssey', '8.256-8.294'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-odyssey', '8.295-8.342'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'ovid-metamorphoses', '14.566-14.608'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'ovid-metamorphoses', '15.807-15.842'),
    ('Aphrodite', 'parentage', 'daughter of Dione', 'homer-iliad', '5.334-5.380'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'hesiod-homeric-hymns', '3.182-3.238'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'hesiod-homeric-hymns', '5.106-5.154'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'hesiod-homeric-hymns', '5.155-5.201'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'hesiod-homeric-hymns', '5.53-5.105'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '14.153-14.196'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '14.197-14.241'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '21.382-21.422'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '23.161-23.191'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '3.340-3.380'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '5.121-5.165'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '5.297-5.333'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '5.334-5.380'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '5.778-5.824'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-odyssey', '8.295-8.342'),
    ('Aphrodite', 'parentage', 'daughter of Zeus (the mighty son of Cronos)', 'homer-odyssey', '8.256-8.294'),
    ('Io', 'parentage', 'Io is the daughter of Inachus', 'ovid-metamorphoses', '1.622-1.667'),
    ('Io', 'parentage', 'child of Iasus', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'child of Inachus', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'child of Inachus', 'ovid-metamorphoses', '1.567-1.600'),
    ('Io', 'parentage', 'child of Inachus', 'ovid-metamorphoses', '1.622-1.667'),
    ('Io', 'parentage', 'child of Piren', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'daughter of Iasus, according to the main narrative', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'daughter of Inachus', 'ovid-metamorphoses', '1.567-1.600'),
    ('Io', 'parentage', 'daughter of Inachus, according to Castor and many tragedians', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'daughter of Piren, according to Hesiod and Acusilaus', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
]

In [ ]:
from datetime import datetime, timezone

from extraction.run_extraction import _claim_key

PROMOTION_LOG_PATH = INGESTION_ROOT / "audit" / "promotion_log.json"
BATCH_LABEL = "p4-review-session"  # <-- EDIT before running: a short label for this session's batch

# Track C4 (DEV-104): a key that matches no row is an error, not a no-op -- refuse to write
# rather than silently skip it. The claim may have been dropped by a re-extraction (see the
# WARNING run_extraction.py prints when that happens) or simply mistyped.
key_to_index = {_claim_key(c): i for i, c in enumerate(candidates)}
unmatched = [k for k in approved_keys if k not in key_to_index]
if unmatched:
    raise SystemExit(
        f"{len(unmatched)} approved key(s) match no row in {CANDIDATES_PATH.name} -- refusing to "
        "write:\n" + "\n".join(f"  {k}" for k in unmatched)
    )

newly_promoted = [k for k in approved_keys if candidates[key_to_index[k]]["trust_tier"] != 1]
for k in approved_keys:
    candidates[key_to_index[k]]["trust_tier"] = 1

CANDIDATES_PATH.write_text(json.dumps(candidates, indent=2, ensure_ascii=False), encoding="utf-8")
total_promoted = sum(1 for c in candidates if c["trust_tier"] == 1)
print(
    f"{len(newly_promoted)} row(s) newly promoted this run "
    f"({len(approved_keys) - len(newly_promoted)} already were); "
    f"{total_promoted} total rows now at trust_tier=1 (out of {len(candidates)})"
)

# Track C5 (DEV-104): per-batch promotion log, appended (never overwritten/replaced) -- lets a
# batch's promotions be identified and reverted as a unit if Track F's compare.py comes back red.
# Only real, new promotions are logged -- re-running an already-promoted key's line (e.g. after
# the C1/C3 migration above) logs nothing, since nothing changed.
if newly_promoted:
    log_entries = json.loads(PROMOTION_LOG_PATH.read_text(encoding="utf-8")) if PROMOTION_LOG_PATH.exists() else []
    log_entries.append(
        {
            "batchLabel": BATCH_LABEL,
            "date": datetime.now(timezone.utc).isoformat(),
            "keys": [list(k) for k in newly_promoted],
            "groupCount": len({(k[0], k[1]) for k in newly_promoted}),
        }
    )
    PROMOTION_LOG_PATH.parent.mkdir(parents=True, exist_ok=True)
    PROMOTION_LOG_PATH.write_text(json.dumps(log_entries, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(
        f"logged to {PROMOTION_LOG_PATH.name}: batch={BATCH_LABEL!r}, {len(newly_promoted)} key(s), "
        f"{log_entries[-1]['groupCount']} group(s)"
    )

## B6 — Guarantee the three floor conflicts

Check which floor conflicts are already covered by promoted (`trust_tier=1`) rows.
Extraction-preferred: only hand-add the ones missing below.

In [26]:
def floor_conflict_covered(subject: str, claim_type: str, min_distinct_values: int = 2) -> bool:
    promoted = [
        c for c in candidates
        if c["trust_tier"] == 1
        and c["subject_name"].strip().lower() == subject.lower()
        and c["claim_type"] == claim_type
    ]
    return len({c["claim_value"] for c in promoted}) >= min_distinct_values


coverage = {
    "Aphrodite parentage": floor_conflict_covered("Aphrodite", "parentage"),
    "Io parentage": floor_conflict_covered("Io", "parentage"),
    "Achilles death": floor_conflict_covered("Achilles", "death"),
}
for name, covered in coverage.items():
    print(f"{name}: {'covered' if covered else 'MISSING -- hand-add below'}")

Aphrodite parentage: covered
Io parentage: covered
Achilles death: covered


## Hand-add any floor conflict extraction missed

**Keep the `claim_type` canonical** (per DEV-018/DEV-020: Achilles' is `death`, not
`slaying`/`death_manner`) so a hand-added row groups with any extracted counterpart and
the exact-match `ConflictLookup` returns both. Run the normalize check first to
confirm the value you're about to use.

In [ ]:
import psycopg2

import config
from extraction.claim_type_normalizer import load_alias_map, normalize

conn = psycopg2.connect(
    host=config.POSTGRES_HOST, port=config.POSTGRES_PORT, dbname=config.POSTGRES_DB,
    user=config.POSTGRES_USER, password=config.POSTGRES_PASSWORD,
)
alias_map = load_alias_map(conn)
conn.close()

print(normalize(alias_map, "slaying"))    # -> "death"
print(normalize(alias_map, "parent_of"))  # -> "parentage"

In [ ]:
# Edit and run only for floor conflicts flagged MISSING above. Example shown for Io
# (Inachus vs Piren, per Apollodorus -- confirm the real passage_ref before running).
hand_added = [
    {
        "subject_name": "Io",
        "claim_type": "parentage",
        "claim_value": "daughter of Inachus",
        "source_id": "apollodorus-bibliotheca",
        "passage_ref": "2.1.3",
        "trust_tier": 1,
    },
    {
        "subject_name": "Io",
        "claim_type": "parentage",
        "claim_value": "daughter of Piren",
        "source_id": "apollodorus-bibliotheca",
        "passage_ref": "2.1.3",
        "trust_tier": 1,
    },
]

candidates.extend(hand_added)
CANDIDATES_PATH.write_text(json.dumps(candidates, indent=2, ensure_ascii=False), encoding="utf-8")
print(f"{len(hand_added)} rows hand-added; {len(candidates)} total rows now in the file")

## Final check

Re-run the B6 coverage cell above after any hand-adds — all three floor conflicts
must show `covered` before handing this off for Track C (`V12`).